crates/core/c8y_api/src/json_c8y.rs

```diff
@@ -236,96 +240,184 @@ impl SourceInfo {
-    ) -> Self {
-        Self {
-            source,
-            severity,
-            alarm_type,
-            time,
-            text,
-            fragments,
+/// Internal representation of clearing an alarm in c8y.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct C8yClearAlarm {
+    /// Alarm type, default is "ThinEdgeAlarm".
+    #[serde(rename = "type")]
+    pub alarm_type: String,
+
+    /// None for main device, Some for child device.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "externalSource")]
+    pub source: Option<SourceInfo>,
+}
+
+impl C8yAlarm {
+    pub fn try_from(
+        alarm: &ThinEdgeAlarm,
+        entity_store: &EntityStore,
+    ) -> Result<Self, C8yAlarmError> {
+        if let Some(entity) = entity_store.get(&alarm.source) {
+            let source = Self::convert_source(entity);
+            let alarm_type = Self::convert_alarm_type(&alarm.alarm_type);
+
+            let c8y_alarm = match alarm.data.as_ref() {
+                None => C8yAlarm::Clear(C8yClearAlarm { alarm_type, source }),
+                Some(tedge_alarm_data) => C8yAlarm::Create(C8yCreateAlarm {
+                    alarm_type: alarm_type.clone(),
+                    source,
+                    severity: C8yCreateAlarm::convert_severity(tedge_alarm_data),
+                    text: C8yCreateAlarm::convert_text(tedge_alarm_data, &alarm_type),
+                    time: C8yCreateAlarm::convert_time(tedge_alarm_data),
+                    fragments: C8yCreateAlarm::convert_extras(tedge_alarm_data),
+                }),
+            };
+            Ok(c8y_alarm)
+        } else {
+            Err(C8yAlarmError::UnsupportedExternalDeviceId(
+                alarm.source.to_string(),
+            ))
         }
     }
-}
 
-impl TryFrom<&ThinEdgeAlarm> for C8yCreateAlarm {
-    type Error = SMCumulocityMapperError;
-
-    fn try_from(alarm: &ThinEdgeAlarm) -> Result<Self, SMCumulocityMapperError> {
-        let severity = alarm.severity.to_string();
-        let alarm_type = alarm.name.to_owned();
-        let text;
-        let time;
-        let fragments;
-
-        match &alarm.to_owned().data {
-            None => {
-                text = alarm_type.clone();
-                time = OffsetDateTime::now_utc();
-                fragments = HashMap::new();
-            }
-            Some(data) => {
-                text = data.text.clone().unwrap_or_else(|| alarm_type.clone());
-                time = data.time.unwrap_or_else(OffsetDateTime::now_utc);
-                fragments = data.alarm_data.clone();
-            }
+    fn convert_source(entity: &EntityMetadata) -> Option<SourceInfo> {
+        match entity.r#type {
+            EntityType::MainDevice => None,
+            EntityType::ChildDevice => Some(make_c8y_source_fragment(&entity.entity_id.clone())),
+            EntityType::Service => None, // TODO!: What is for service?
```

Comment:
A service is addressed the same way as for a child-device.

Reply:
Done with 816a8cb06525fbaef157ef25690ba983ed647949